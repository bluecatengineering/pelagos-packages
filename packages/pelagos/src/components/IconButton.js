import {forwardRef} from 'react';
import PropTypes from 'prop-types';

import setRefs from '../functions/setRefs';
import useTooltip from '../hooks/useTooltip';

import SvgIcon from './SvgIcon';
import './IconButton.less';

/** An icon button. */
const IconButton = forwardRef(
	({id, icon, className, size, type, tooltipText, tooltipPlacement, overlay, disabled, onClick, ...props}, ref) => {
		const tooltip = useTooltip(tooltipText, tooltipPlacement);
		const refs = ref ? setRefs(ref, tooltip) : tooltip;
		return disabled ? (
			<span
				{...props}
				id={id}
				className={`IconButton IconButton--${size} IconButton--${type}${className ? ` ${className}` : ''}`}
				role="button"
				aria-disabled="true"
				ref={refs}>
				<SvgIcon icon={icon} />
				{overlay}
			</span>
		) : (
			<button
				{...props}
				id={id}
				className={`IconButton IconButton--${size} IconButton--${type}${className ? ` ${className}` : ''}`}
				type="button"
				ref={refs}
				onClick={onClick}>
				<SvgIcon icon={icon} />
				{overlay}
			</button>
		);
	}
);

IconButton.displayName = 'IconButton';

IconButton.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The object representing the icon. (using FontAwesome, etc.) */
	icon: PropTypes.object.isRequired,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The component size. */
	size: PropTypes.oneOf(['medium', 'large']),
	/** The button type. */
	type: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'ghost']),
	/** The tooltip text to display. */
	tooltipText: PropTypes.string,
	/** The placement of the tooltip relative to the button. */
	tooltipPlacement: PropTypes.oneOf(['left', 'right', 'top', 'bottom']),
	/** An overlay element. */
	overlay: PropTypes.node,
	/** Whether the button is disabled. */
	disabled: PropTypes.bool,
	/** Function invoked when the button is clicked. */
	onClick: PropTypes.func,
};

IconButton.defaultProps = {
	size: 'medium',
	type: 'ghost',
	tooltipPlacement: 'right',
};

export default IconButton;
