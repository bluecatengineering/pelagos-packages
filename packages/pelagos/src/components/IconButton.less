@import '../../less/spacing';
@import '../../less/utils';

@layer pelagos {
	.IconButton {
		@focus-visible();
		display: inline-flex;
		align-items: center;
		justify-content: center;
		border-radius: 3px;
		cursor: pointer;
		position: relative;

		&[aria-expanded='true'],
		&[aria-expanded='true']:focus {
			border-radius: initial;
			box-shadow: var(--shadow-04);
		}

		&--medium {
			font-size: 16px;
			width: 32px;
			height: 32px;
		}

		&--large {
			font-size: 24px;
			width: 48px;
			height: 48px;
		}

		&--primary {
			@primary-button-effects();
			color: var(--text-on-color);
			background-color: var(--button-primary);
			border: 1px solid transparent;

			&[aria-expanded='true'] {
				background-color: var(--button-primary-active);
			}

			&[aria-disabled='true'] {
				color: var(--text-on-color-disabled);
				background-color: var(--button-disabled);
			}
		}

		&--secondary {
			@secondary-button-effects();
			color: var(--text-on-color);
			background-color: var(--button-secondary);
			border: 1px solid transparent;

			&[aria-disabled='true'] {
				color: var(--text-on-color-disabled);
				background-color: var(--button-disabled);
			}
		}

		&--tertiary {
			@tertiary-button-effects();
			color: var(--button-tertiary);
			background-color: transparent;
			border: 1px solid var(--button-tertiary);

			&[aria-expanded='true'] {
				background-color: var(--button-tertiary-active);
			}

			&[aria-disabled='true'] {
				color: var(--text-on-color-disabled);
				background-color: transparent;
				border-color: var(--button-disabled);
			}
		}

		&--ghost {
			@layer-hover-effects();
			color: var(--link-primary);
			background-color: transparent;
			border: 1px solid transparent;

			&[aria-expanded='true'] {
				background-color: var(--layer);
			}

			&[aria-disabled='true'] {
				color: var(--icon-disabled);
				background-color: transparent;
			}
		}

		&[aria-disabled='true'] {
			cursor: not-allowed;
		}

		&:focus-visible {
			box-shadow: inset 0 0 0 2px var(--focus-inset);
		}

		@media print {
			display: none;
		}
	}
}
