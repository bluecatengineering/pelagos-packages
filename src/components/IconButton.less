@import '../../less/colors';

.IconButton {
	display: inline-block;
	font-size: 19px;
	cursor: pointer;
	padding: 2px;
	color: @azure;
	transition: color 0.3s ease-out;
	vertical-align: -4px;
}

.IconButton--large {
	font-size: 31px;
}

.IconButton:hover {
	color: @sea;
}

.IconButton:active {
	color: @ocean;
}

.IconButton:focus {
	outline: 2px solid @bondi;
	outline-offset: 0;
}

.IconButton[aria-disabled='true'],
.IconButton[aria-disabled='true']:hover,
.IconButton[aria-disabled='true']:active {
	color: @battleship;
	cursor: not-allowed;
}
