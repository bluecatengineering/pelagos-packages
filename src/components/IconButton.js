import PropTypes from 'prop-types';

import useTooltip from '../hooks/useTooltip';

import SvgIcon from './SvgIcon';
import './IconButton.less';

/** An icon button. */
const IconButton = ({id, icon, className, size, type, tooltipText, tooltipPlacement, disabled, onClick, ...props}) => {
	const tooltip = useTooltip(tooltipText, tooltipPlacement);
	return disabled ? (
		<span
			{...props}
			id={id}
			className={`IconButton IconButton--${size} IconButton--${type}${className ? ` ${className}` : ''}`}
			aria-disabled="true"
			ref={tooltip}
		>
			<SvgIcon icon={icon} />
		</span>
	) : (
		<button
			{...props}
			id={id}
			className={`IconButton IconButton--${size} IconButton--${type}${className ? ` ${className}` : ''}`}
			type="button"
			ref={tooltip}
			onClick={onClick}
		>
			<SvgIcon icon={icon} />
		</button>
	);
};

IconButton.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The object representing the icon. (using FontAwesome, etc.) */
	icon: PropTypes.object.isRequired,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The component size. */
	size: PropTypes.oneOf(['medium', 'large']),
	/** The button type. */
	type: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'ghost']),
	/** The tooltip text to display. */
	tooltipText: PropTypes.string,
	/** The placement of the tooltip relative to the button. */
	tooltipPlacement: PropTypes.oneOf(['left', 'right', 'top', 'bottom']),
	/** Whether the button is disabled. */
	disabled: PropTypes.bool,
	/** Function invoked when the button is clicked. */
	onClick: PropTypes.func,
};

IconButton.defaultProps = {
	size: 'medium',
	type: 'ghost',
	tooltipPlacement: 'right',
};

export default IconButton;
