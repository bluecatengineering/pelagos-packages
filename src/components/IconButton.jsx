import React from 'react';
import PropTypes from 'prop-types';

import useButtonKeyHandler from '../hooks/useButtonKeyHandler';

import Tooltip from './Tooltip';
import OverlayTrigger from './OverlayTrigger';
import SvgIcon from './SvgIcon';
import './IconButton.less';

const IconButton = ({id, icon, className, tooltipText, large, disabled, onClick, ...props}) => {
	const keyHandker = useButtonKeyHandler(onClick);
	const button = (
		<span
			{...props}
			id={id}
			role="button"
			aria-disabled={disabled}
			tabIndex={disabled ? -1 : 0}
			className={'IconButton' + (className ? ' ' + className : '') + (large ? ' IconButton--large' : '')}
			onClick={disabled ? null : onClick}
			onKeyDown={disabled ? null : keyHandker}>
			<SvgIcon icon={icon} />
		</span>
	);
	if (!tooltipText) {
		return button;
	}

	const tooltipId = id + '-tooltip';
	const tooltip = <Tooltip id={tooltipId}>{tooltipText}</Tooltip>;
	return (
		<OverlayTrigger placement={large ? 'left' : 'right'} overlay={tooltip}>
			{button}
		</OverlayTrigger>
	);
};

IconButton.propTypes = {
	id: PropTypes.string,
	icon: PropTypes.object.isRequired,
	className: PropTypes.string,
	tooltipText: PropTypes.string,
	large: PropTypes.bool,
	disabled: PropTypes.bool,
	onClick: PropTypes.func,
};

export default IconButton;
