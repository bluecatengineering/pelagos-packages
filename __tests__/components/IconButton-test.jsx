import React from 'react';
import {shallow} from 'enzyme';

import IconButton from '../../src/components/IconButton';

jest.unmock('../../src/components/IconButton');

describe('IconButton', () => {
	describe('rendering', () => {
		it('renders the expected elements', () => {
			const wrapper = shallow(
				<IconButton componentId="test" icon={{}} className="TestClass" ariaLabel="Test" tooltipText="This is a test" />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders the expected elements when className is not set', () => {
			const wrapper = shallow(
				<IconButton componentId="test" icon={{}} ariaLabel="Test" tooltipText="This is a test" />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders the expected elements when tooltipText is not set', () => {
			const wrapper = shallow(<IconButton componentId="test" icon={{}} className="TestClass" ariaLabel="Test" />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders the expected elements if disabled is true', () => {
			const wrapper = shallow(
				<IconButton componentId="test" icon={{}} ariaLabel="Test" tooltipText="This is a test" disabled={true} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders the expected elements if large is true', () => {
			const wrapper = shallow(
				<IconButton componentId="test" icon={{}} ariaLabel="Test" tooltipText="This is a test" large={true} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onClick when the button is clicked', () => {
			const onClick = jest.fn();
			const wrapper = shallow(<IconButton componentId="test" icon={{}} onClick={onClick} />);
			wrapper.find('[data-bcn-id="btn-test"]').simulate('click');
			expect(onClick).toHaveBeenCalled();
		});

		it('does not call onClick when the button is disabled and clicked', () => {
			const onClick = jest.fn();
			const wrapper = shallow(<IconButton componentId="test" icon={{}} disabled={true} onClick={onClick} />);
			wrapper.find('[data-bcn-id="btn-test"]').simulate('click');
			expect(onClick).not.toHaveBeenCalled();
		});

		it('calls onClick when enter is pressed', () => {
			const onClick = jest.fn();
			const event = {keyCode: 13, preventDefault: jest.fn()};
			const wrapper = shallow(<IconButton componentId="test" icon={{}} onClick={onClick} />);
			wrapper.find('[data-bcn-id="btn-test"]').simulate('keydown', event);
			expect(onClick.mock.calls).toEqual([[event]]);
			expect(event.preventDefault).toHaveBeenCalled();
		});

		it('calls onClick when space is pressed', () => {
			const onClick = jest.fn();
			const event = {keyCode: 32, preventDefault: jest.fn()};
			const wrapper = shallow(<IconButton componentId="test" icon={{}} onClick={onClick} />);
			wrapper.find('[data-bcn-id="btn-test"]').simulate('keydown', event);
			expect(onClick.mock.calls).toEqual([[event]]);
			expect(event.preventDefault).toHaveBeenCalled();
		});

		it('does not call onClick when any other key is pressed', () => {
			const onClick = jest.fn();
			const event = {keyCode: 65, preventDefault: jest.fn()};
			const wrapper = shallow(<IconButton componentId="test" icon={{}} onClick={onClick} />);
			wrapper.find('[data-bcn-id="btn-test"]').simulate('keydown', event);
			expect(onClick).not.toHaveBeenCalled();
			expect(event.preventDefault).not.toHaveBeenCalled();
		});

		it('does not call onClick when the button is disabled and a key is pressed', () => {
			const onClick = jest.fn();
			const wrapper = shallow(<IconButton componentId="test" icon={{}} disabled={true} onClick={onClick} />);
			wrapper.find('[data-bcn-id="btn-test"]').simulate('keydown');
			expect(onClick).not.toHaveBeenCalled();
		});
	});
});
