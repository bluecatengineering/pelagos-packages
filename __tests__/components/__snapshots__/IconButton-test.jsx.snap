// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton rendering renders the expected elements 1`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="test-tooltip"
    >
      This is a test
    </Tooltip>
  }
  placement="right"
>
  <span
    aria-label="Test"
    className="IconButton TestClass"
    id="test"
    role="button"
    tabIndex={0}
  >
    <SvgIcon
      icon={Object {}}
    />
  </span>
</OverlayTrigger>
`;

exports[`IconButton rendering renders the expected elements if disabled is true 1`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="test-tooltip"
    >
      This is a test
    </Tooltip>
  }
  placement="right"
>
  <span
    aria-disabled={true}
    aria-label="Test"
    className="IconButton"
    id="test"
    onClick={null}
    onKeyDown={null}
    role="button"
    tabIndex={-1}
  >
    <SvgIcon
      icon={Object {}}
    />
  </span>
</OverlayTrigger>
`;

exports[`IconButton rendering renders the expected elements if large is true 1`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="test-tooltip"
    >
      This is a test
    </Tooltip>
  }
  placement="left"
>
  <span
    aria-label="Test"
    className="IconButton IconButton--large"
    id="test"
    role="button"
    tabIndex={0}
  >
    <SvgIcon
      icon={Object {}}
    />
  </span>
</OverlayTrigger>
`;

exports[`IconButton rendering renders the expected elements when className is not set 1`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="test-tooltip"
    >
      This is a test
    </Tooltip>
  }
  placement="right"
>
  <span
    aria-label="Test"
    className="IconButton"
    id="test"
    role="button"
    tabIndex={0}
  >
    <SvgIcon
      icon={Object {}}
    />
  </span>
</OverlayTrigger>
`;

exports[`IconButton rendering renders the expected elements when tooltipText is not set 1`] = `
<span
  aria-label="Test"
  className="IconButton TestClass"
  id="test"
  role="button"
  tabIndex={0}
>
  <SvgIcon
    icon={Object {}}
  />
</span>
`;
